File: day19/p1.py
# Answer: 269

lines = open("input.txt").read().splitlines()

patterns = set(lines[0].split(", "))
max_len = max(map(len, patterns))
cache = {}

def can_obtain(design):
    if design == "":
        return True
    if design in cache:
        return cache[design]
    for i in range(min(len(design), max_len) + 1):
        if design[:i] in patterns and can_obtain(design[i:]):
            cache[design] = True
            return True
    return False

print(sum(can_obtain(design) for design in lines[2:]))

File: day19/p1.js
const fs = require('fs');

// Read input from the file
const lines = fs.readFileSync('input.txt', 'utf-8').trim().split("\n");
const units = lines[0].split(", ");
const designs = lines.slice(2);

// Function to check if a design can be made using the units
function possible(design) {
    const n = design.length;
    const dp = Array(n).fill(false); // Initialize dp array with false values

    for (let i = 0; i < n; i++) {
        if (units.includes(design.slice(0, i + 1))) {
            dp[i] = true;
            continue;
        }

        for (const unit of units) {
            if (i - unit.length + 1 >= 0 && design.slice(i - unit.length + 1, i + 1) === unit && dp[i - unit.length]) {
                dp[i] = true;
                break;
            }
        }
    }

    return dp[n - 1];
}

let ans = 0;
for (const design of designs) {
    if (possible(design)) {
        console.log(design);
        ans++;
    }
}

console.log(ans);
